### PR TITLE
[Common] tensor data struct

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Now tensor-transform and tensor-if use common data struct to handle data.
Add new struct and common functions to handle tensor type and value.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_data.c [hunk 1]
```diff
@@ -0,0 +1,296 @@
+/**
+ * SPDX-License-Identifier: LGPL-2.1-only
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_data.c:2]: Separate SPDX line. Put it at LINE-1.

**Review by myungjoo — APPROVED** [review]
LGTM other than SPDX styling.

**Review by gichan-jang — APPROVED** [review]
LGTM!

File: gst/nnstreamer/tensor_data.c (hunk 1 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c]: updated

File: gst/nnstreamer/tensor_data.c [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    is_float = (td->type == _NNS_FLOAT32 || td->type == _NNS_FLOAT64);
+
+    switch (type) {
+      case _NNS_INT32:
+        td_typecast (td, int32_t);
+        break;
+      case _NNS_UINT32:
+        if (is_float) {
+          td_typecast (td, int32_t);
+          td->type = _NNS_INT32;
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_data.c:188]: I wonder if this is redundant as `td_typecast` sets the new type. (below as well)

Comment (zhoonit) [inline gst/nnstreamer/tensor_data.c:188]: Maybe this? line 233 overwrites the type
```suggestion
          type = _NNS_INT32;
```

File: gst/nnstreamer/tensor_data.c [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  tensor_data_s td;
+
+  g_return_val_if_fail (input != NULL, FALSE);
+  g_return_val_if_fail (output != NULL, FALSE);
+  g_return_val_if_fail (in_type != _NNS_END, FALSE);
+  g_return_val_if_fail (out_type != _NNS_END, FALSE);
+
+  gst_tensor_data_set (&td, in_type, input);
+  gst_tensor_data_typecast (&td, out_type);
+  gst_tensor_data_get (&td, output);
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_data.c:260]: return false if one of them fails?

File: gst/nnstreamer/tensor_data.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+extern gboolean
+gst_tensor_data_set (tensor_data_s * td, tensor_type type, gpointer value);
+
+/**
+ * @brief Get tensor element value.
+ * @param td struct for tensor data
+ * @param value pointer of tensor element value
+ * @return TRUE if no error
+ */
+extern  gboolean
```
Comment (zhoonit) [inline gst/nnstreamer/tensor_data.h:46]:
```suggestion
extern gboolean
```

**Review by anyj0527 — APPROVED** [review]

File: gst/nnstreamer/tensor_data.c (hunk 2 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c:188]: no, changing float > int32 > uint32 and float > uint32 is different.

File: gst/nnstreamer/tensor_data.c (hunk 3 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c:260]: IMO, these lines are always true with the conditions in line 253~256

File: gst/nnstreamer/tensor_data.h (hunk 4 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.h:46]: I will update this line.

File: gst/nnstreamer/tensor_data.c (hunk 2 above)
Comment (zhoonit) [reply gst/nnstreamer/tensor_data.c:188]: Ah I see :) Thank you for the explanation!

**Review by zhoonit — APPROVED** [review]
LGTM 👍

File: gst/nnstreamer/tensor_data.c (hunk 3 above)
Comment (zhoonit) [reply gst/nnstreamer/tensor_data.c:260]: I see. Thank you for the clarification :)

File: gst/nnstreamer/tensor_data.c [hunk 5]
```diff
@@ -0,0 +1,296 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file	tensor_data.c
+ * @date	10 Mar 2021
+ * @brief	Internal functions to handle various tensor type and value.
+ * @see		http://github.com/nnstreamer/nnstreamer
```
Comment (wooksong) [inline gst/nnstreamer/tensor_data.c:8]: Mixed tabs and spaces in the indentation of this line.

File: gst/nnstreamer/tensor_data.c [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+ * @param type tensor type to be transformed
+ * @return TRUE if no error
+ */
+gboolean
+gst_tensor_data_typecast (tensor_data_s * td, tensor_type type)
+{
+  gboolean is_float;
+
+  g_return_val_if_fail (td != NULL, FALSE);
+  g_return_val_if_fail (type != _NNS_END, FALSE);
```
Comment (wooksong) [inline gst/nnstreamer/tensor_data.c:175]: Is it necessary? The ```default``` case in the switch block below can handle all the cases of invalid ```type```.

File: gst/nnstreamer/tensor_transform/tensor_transform.h [hunk 7]
```diff
@@ -35,6 +35,7 @@
 #include <gst/gst.h>
 #include <gst/base/gstbasetransform.h>
 #include <tensor_common.h>
+#include <tensor_data.h>
```
Comment (wooksong) [inline gst/nnstreamer/tensor_transform/tensor_transform.h:38]: Do we export this header?

File: gst/nnstreamer/tensor_data.h (hunk 4 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.h]: updated

File: gst/nnstreamer/tensor_data.c (hunk 5 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c:8]: In code view, this is correct.
https://github.com/jaeyun-jung/nnstreamer/blob/0c3db1f73e0f5a56fd5085f0d4c0a9f2584656ea/gst/nnstreamer/tensor_data.c#L8

File: gst/nnstreamer/tensor_transform/tensor_transform.h (hunk 7 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_transform/tensor_transform.h:38]: export for devel pkg? no, this header file is from different include-path in build script.

File: gst/nnstreamer/tensor_data.c (hunk 6 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c:175]: updated

File: gst/nnstreamer/tensor_data.c (hunk 5 above)
Comment (wooksong) [reply gst/nnstreamer/tensor_data.c:8]: @jaeyun-jung These are tabs. 

image

File: gst/nnstreamer/tensor_transform/tensor_transform.h (hunk 7 above)
Comment (wooksong) [reply gst/nnstreamer/tensor_transform/tensor_transform.h:38]: IMHO, although the directory where this header is placed is provided with ```-I``` option, it is better to directly indicate the user header path for better compile performance (maybe old-school).

```c
#include "../tensor_data.h"
```

Whatever you choose, LGTM here.

File: gst/nnstreamer/tensor_data.c (hunk 5 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_data.c]: changed to single tab for the tag `@see / @bug`

**Review by wooksong — APPROVED** [review]

**Pull request merged**